routes/profile.js

```diff
@@ -52,54 +52,89 @@ async function getProfileData(profileName, geoids, compare, db) {
-  profileDF = profileDF.join(
-    new DataIngestor(
-      previousProfileData,
-      profileName,
-      isAggregate,
-      /* is previous */ true,
-    ).processRaw(),
-    'variable',
-  );
-
-  // join with compare profile data Dataframe
-  profileDF = profileDF.join(
-    // 'compare' profiles are always NOT aggregate; they are comprised of a single geoid
-    new DataIngestor(
-      compareProfileData,
-      profileName,
-      /* is aggregate */ false,
-      /* is previous */ false,
-      /* is comparison */ true,
-    ).processRaw(),
-    'variable',
-  );
-
-  // convert combined Dataframe to array of objects
-  // (change & difference calculations requires ad-hoc property addition,
-  // which is *MUCH* easier with objects than Dataframe rows)
-  // TODO: confirm using Dataframe is faster than just iterating over objects...
-  return profileDF
-    .toCollection()
+  new DataProcessor(profileData, profileName, isAggregate).process();
+  new DataProcessor(previousProfileData, profileName, isAggregate, /* isPrevious */ true).process();
+  new DataProcessor(compareProfileData, profileName, /* isAggregate */ false).process();
+
+  join(profileData, previousProfileData, compareProfileData);
+
+  
+  profileData
     .map((row) => {
       const varConfig = find(specialCalculationConfigs[profileName], ['variable', row.variable]) || {};
       doChangeCalculations(row, varConfig);
       doDifferenceCalculations(row);
-      removePercentsFromSpecialVars(row, varConfig);
-      return row;
-    });
+  });
+  return profileData;
+}
+
+/* 
+ * Joins profile, previousProfile, and compareProfile row objects,
+ * prepending key names with appropriate prefixes before combining.
+ * (previous and compare, respectively).
+ * @param{Object[]} profile - Array of profile row objects
+ * @param{Object[]} previous - Array of previous profile row objects
+ * @param{Object[]} compare - Array of compare profile row objects
+ */
+function join(profile, previous, compare) {
+  const valueKeys = getValueKeys(Object.keys(profile[0]));
+
+  profile.sort(profileSort);   
+  previous.sort(profileSort);
+  compare.sort(profileSort);
+
+  for(let i = 0; i < profile.length; i++) {
+    const row = profile[i];
+    const previousRow = previous[i];
+    const compareRow = compare[i];
+
+    addValuesToRow(row, previousRow, 'previous', valueKeys);
+    addValuesToRow(row, compareRow, 'comparison', valueKeys);
+  }
+}
+
+/* 
+ * Returns the keys for actual values in a row object,
+ * filtering out all metadata, as these fields do not need to
+ * be preserved from the previous and compare row objects.
+ * @param{String[]} allKeys - Array containing all of keys in the row object
+ */
+function getValueKeys(allKeys) {
+  return allKeys.filter((v) => {
+    return v !== 'id'
+    && v !== 'variable'
+    && v !== 'variablename'
+    && v !== 'base'
+    && v !== 'category'
+    && v !== 'profile'
+    && v !== 'base_sum'
+    && v !== 'base_m';
+  });
+}
```

Comment:
What about writing this way? Is this more confusing? ->

```javascript
function getValueKeys(allKeys) {
  return allKeys.filter((v) => {
    return !['id','variable','variablename','base','category','profile','base_sum','base_m'].includes(v)
  });
}
```

Reply:
nope u know i love this!